transport/httptransport/http_transport.go

```diff
@@ -121,41 +166,72 @@ func (t *transfer) emitEvent(ctx context.Context, evt types.TransportEvent, id u
 }
 
 func (t *transfer) execute(ctx context.Context) error {
-	// construct request
-	req, err := http.NewRequest("GET", t.tInfo.URL, nil)
-	if err != nil {
-		return fmt.Errorf("failed to create http req: %w", err)
-	}
 
-	// add range req to start reading from the last byte we have in the output file
-	req.Header.Set("Range", fmt.Sprintf("bytes=%d-", t.nBytesReceived))
-	// init the request with the transfer context
-	req = req.WithContext(ctx)
-	// open output file in append-only mode for writing
-	of, err := os.OpenFile(t.dealInfo.OutputFile, os.O_APPEND|os.O_WRONLY, 0644)
-	if err != nil {
-		return fmt.Errorf("failed to open output file: %w", err)
-	}
-	defer of.Close()
+	for {
+		// construct request
+		req, err := http.NewRequest("GET", t.tInfo.URL, nil)
+		if err != nil {
+			return fmt.Errorf("failed to create http req: %w", err)
+		}
 
-	// start the http transfer
-	remaining := t.dealInfo.DealSize - t.nBytesReceived
-	if err := t.doHttp(ctx, req, of, remaining); err != nil {
-		// do not resume transfer if context has been cancelled
-		if ctx.Err() == context.Canceled || ctx.Err() == context.DeadlineExceeded {
-			return fmt.Errorf("transfer context err: %w", ctx.Err())
+		st, err := os.Stat(t.dealInfo.OutputFile)
+		if err != nil {
+			return fmt.Errorf("failed to stat output file: %w", err)
+		}
+		t.nBytesReceived = st.Size()
+
+		// add range req to start reading from the last byte we have in the output file
+		req.Header.Set("Range", fmt.Sprintf("bytes=%d-", t.nBytesReceived))
+		// init the request with the transfer context
+		req = req.WithContext(ctx)
+		// open output file in append-only mode for writing
+		of, err := os.OpenFile(t.dealInfo.OutputFile, os.O_APPEND|os.O_WRONLY, 0644)
+		if err != nil {
+			return fmt.Errorf("failed to open output file: %w", err)
+		}
+		defer of.Close()
+
+		// start the http transfer
+		remaining := t.dealInfo.DealSize - t.nBytesReceived
+		if err = t.doHttp(ctx, req, of, remaining); err == nil {
+			// if there's no error, transfer was successful
+			break
+		}
+		_ = of.Close()
```

Comment:
Do we need of.Close() here? It seems like it's also being called in the defer statement above on line 192
Also that defer will only get called when the function exits, I'm not sure if that's what we want?
It may be easier to do with a defer inside a function called from within the for loop.

Reply:
Hey, we need to close before doing the resumption just to be safe.

Reply:
I'm not sure stating on an already open file will work out well.